sdk/core/azure-core/src/main/java/com/azure/core/util/polling/SyncPoller.java

```diff
@@ -66,6 +66,34 @@ public interface SyncPoller<T, U> {
      */
     PollResponse<T> waitForCompletion(Duration timeout);
 
+    /**
+     * Wait for polling to complete with a timeout. The polling is considered complete based on status defined in
+     * {@link LongRunningOperationStatus} or if one of the timeouts expire.
+     * <p>
+     * The {@code operationTimeout} is applied to each individual poll operation to time it out if the polling operation
+     * takes too long. If a poll operation times out a {@link RuntimeException} wrapping a {@link TimeoutException} will
+     * be thrown.
+     * <p>
+     * The {@code waitTimeout} is used to determine when the wait for should stop. If polling doesn't reach a completion
+     * state before the {@code waitTimeout} elapses a {@link RuntimeException} wrapping a {@link TimeoutException} will
+     * be thrown.
+     * <p>
+     * If this method isn't overridden by the implementation then this method is effectively equivalent to calling
+     * {@link #waitForCompletion(Duration) waitForCompletion(waitTimeout)}.
+     *
+     * @param operationTimeout The duration to wait for each individual poll operation to complete.
+     * @param waitTimeout The duration to wait for polling completion.
+     * @return The final poll response.
+     * @throws NullPointerException If {@code operationTimeout} or {@code waitTimeout} is null.
+     * @throws IllegalArgumentException If {@code operationTimeout} or {@code waitTimeout} is zero or negative or if
+     * {@code operationTimeout} is greater than {@code waitTimeout}.
+     * @throws RuntimeException If polling doesn't complete before the {@code waitTimeout} elapses.
+     * ({@link RuntimeException#getCause()} should be a {@link TimeoutException}).
+     */
+    default PollResponse<T> waitForCompletion(Duration operationTimeout, Duration waitTimeout) {
```

Comment:
I am not sure we need this overload. Polling operation is a single `GET` operation usually. So, this should work like any other GET operation on the client where the timeout is either configured at the HTTP client level or through a TimeoutPolicy. 

If the HTTP client's response timeout is set to 10 seconds and if `waitForCompletion()` is called with operationTimeout set to 60 seconds, I think the request will still timeout after 10 seconds (if retry max attempts set to 0).

So, to configure polling timeout, we should just let that be a client level configuration like all single-request operations.

Reply:
While I agree there are other mechanisms to handle this in the SDKs already, I don't think they're the best general solution. Setting specific timeouts in the HTTP client or using a TimeoutException in creating the client can affect all operations using the HTTP client or HTTP pipeline. These could cause issues if the service provides APIs that may take a while to complete service side and can only be configured once during the lifetime of the client. And as you alluded to, the behavior of those mechanisms can change drastically based on retry policies in the pipeline.

To me, this is a much simpler, clearer, and variable solution. You don't need to dig into the depths of how to configure a SDK client, HTTP client, or HTTP pipeline, and you can change behavior between every call to the API. Also, this gives us a forward path on deprecating the more confusing single timeout APIs which conflate per-operation and overall wait timeouts which makes them very inflexible.